## En nuestra empresa, el equipo de ciencia del dato nos solicita que analicemos y preparemos un texto que luego ellos usarán para entrenar un modelo de Deep Learning.

- El texto nos llega en formato **string**
- Necesitaremos aislar las palabras que lo forman para poder manipularlas.
- Una vez extraídas las palabras, nos piden lo siguiente:

    * Cuántas palabras componen el texto.
    * Cuántas veces se repiten las palabras dentro del texto, en promedio y por palabra.
    * Almacenar el cómputo de veces que se repite cada palabra en orden alfabético.
    * Cuántas veces se repite la palabra que más se repite y cuántas veces se repite la palabra que menos se repite.


https://pypi.org/project/lorem/

In [1]:
!pip install lorem

In [6]:
import lorem

In [8]:
t = lorem.text()
t

'Numquam velit magnam quaerat numquam voluptatem voluptatem. Velit eius neque adipisci modi. Dolorem quiquia non porro est ut amet. Eius non etincidunt quisquam quiquia est labore. Sed dolorem adipisci non aliquam. Est modi non magnam numquam adipisci. Magnam neque est voluptatem modi est eius. Dolor porro aliquam etincidunt sit quisquam.\n\nIpsum dolorem dolore ut ut dolor. Magnam consectetur dolor quiquia dolorem porro sed. Etincidunt amet tempora adipisci porro neque dolore sit. Labore aliquam dolor adipisci quaerat dolorem voluptatem. Tempora numquam velit modi numquam. Modi etincidunt velit quisquam quiquia.\n\nSit etincidunt modi quiquia. Quisquam porro sit ipsum. Tempora velit porro modi. Eius eius quaerat numquam adipisci velit eius. Sed ipsum eius ipsum quiquia quaerat quisquam. Quaerat consectetur velit non magnam. Quaerat consectetur velit velit eius numquam quaerat dolor. Sed amet magnam modi. Sit velit quisquam adipisci dolorem aliquam velit. Ipsum non ipsum quisquam amet 

1. Para aislar las palabras, vemos que no solo tenemos palabras sino también caracteres especiales:

In [9]:
t_limpio = t.replace ('.','')
t_limpio 

'Numquam velit magnam quaerat numquam voluptatem voluptatem Velit eius neque adipisci modi Dolorem quiquia non porro est ut amet Eius non etincidunt quisquam quiquia est labore Sed dolorem adipisci non aliquam Est modi non magnam numquam adipisci Magnam neque est voluptatem modi est eius Dolor porro aliquam etincidunt sit quisquam\n\nIpsum dolorem dolore ut ut dolor Magnam consectetur dolor quiquia dolorem porro sed Etincidunt amet tempora adipisci porro neque dolore sit Labore aliquam dolor adipisci quaerat dolorem voluptatem Tempora numquam velit modi numquam Modi etincidunt velit quisquam quiquia\n\nSit etincidunt modi quiquia Quisquam porro sit ipsum Tempora velit porro modi Eius eius quaerat numquam adipisci velit eius Sed ipsum eius ipsum quiquia quaerat quisquam Quaerat consectetur velit non magnam Quaerat consectetur velit velit eius numquam quaerat dolor Sed amet magnam modi Sit velit quisquam adipisci dolorem aliquam velit Ipsum non ipsum quisquam amet amet non\n\nVelit modi 

Otra manera es hacerlo con la biblioteca string

In [10]:
import string
string.punctuation

'!"#$%&\'()*+,-./:;<=>?@[\\]^_`{|}~'

Si necesitamos agregar mas caracteres a string.punctuation, lo hacemos de la siguiente manera:

In [11]:
print(string.punctuation, 'Clase: ', type(string.punctuation))

!"#$%&'()*+,-./:;<=>?@[\]^_`{|}~ Clase:  <class 'str'>


In [14]:
string.punctuation = string.punctuation + '¡'
print(string.punctuation)

!"#$%&'()*+,-./:;<=>?@[\]^_`{|}~¡¡


In [15]:
tabla_traduccion = str.maketrans('','',string.punctuation)
texto_limpio = t.translate(tabla_traduccion)
print(texto_limpio)

Numquam velit magnam quaerat numquam voluptatem voluptatem Velit eius neque adipisci modi Dolorem quiquia non porro est ut amet Eius non etincidunt quisquam quiquia est labore Sed dolorem adipisci non aliquam Est modi non magnam numquam adipisci Magnam neque est voluptatem modi est eius Dolor porro aliquam etincidunt sit quisquam

Ipsum dolorem dolore ut ut dolor Magnam consectetur dolor quiquia dolorem porro sed Etincidunt amet tempora adipisci porro neque dolore sit Labore aliquam dolor adipisci quaerat dolorem voluptatem Tempora numquam velit modi numquam Modi etincidunt velit quisquam quiquia

Sit etincidunt modi quiquia Quisquam porro sit ipsum Tempora velit porro modi Eius eius quaerat numquam adipisci velit eius Sed ipsum eius ipsum quiquia quaerat quisquam Quaerat consectetur velit non magnam Quaerat consectetur velit velit eius numquam quaerat dolor Sed amet magnam modi Sit velit quisquam adipisci dolorem aliquam velit Ipsum non ipsum quisquam amet amet non

Velit modi consect

2. Poner todas las palabras en minúscula 

In [16]:
texto_limpio = texto_limpio.lower()
texto_limpio

'numquam velit magnam quaerat numquam voluptatem voluptatem velit eius neque adipisci modi dolorem quiquia non porro est ut amet eius non etincidunt quisquam quiquia est labore sed dolorem adipisci non aliquam est modi non magnam numquam adipisci magnam neque est voluptatem modi est eius dolor porro aliquam etincidunt sit quisquam\n\nipsum dolorem dolore ut ut dolor magnam consectetur dolor quiquia dolorem porro sed etincidunt amet tempora adipisci porro neque dolore sit labore aliquam dolor adipisci quaerat dolorem voluptatem tempora numquam velit modi numquam modi etincidunt velit quisquam quiquia\n\nsit etincidunt modi quiquia quisquam porro sit ipsum tempora velit porro modi eius eius quaerat numquam adipisci velit eius sed ipsum eius ipsum quiquia quaerat quisquam quaerat consectetur velit non magnam quaerat consectetur velit velit eius numquam quaerat dolor sed amet magnam modi sit velit quisquam adipisci dolorem aliquam velit ipsum non ipsum quisquam amet amet non\n\nvelit modi 

3. Extraemos las palabras del texto, para eso primero separamos el texto en palabras

In [17]:
texto_limpio.split()

['numquam',
 'velit',
 'magnam',
 'quaerat',
 'numquam',
 'voluptatem',
 'voluptatem',
 'velit',
 'eius',
 'neque',
 'adipisci',
 'modi',
 'dolorem',
 'quiquia',
 'non',
 'porro',
 'est',
 'ut',
 'amet',
 'eius',
 'non',
 'etincidunt',
 'quisquam',
 'quiquia',
 'est',
 'labore',
 'sed',
 'dolorem',
 'adipisci',
 'non',
 'aliquam',
 'est',
 'modi',
 'non',
 'magnam',
 'numquam',
 'adipisci',
 'magnam',
 'neque',
 'est',
 'voluptatem',
 'modi',
 'est',
 'eius',
 'dolor',
 'porro',
 'aliquam',
 'etincidunt',
 'sit',
 'quisquam',
 'ipsum',
 'dolorem',
 'dolore',
 'ut',
 'ut',
 'dolor',
 'magnam',
 'consectetur',
 'dolor',
 'quiquia',
 'dolorem',
 'porro',
 'sed',
 'etincidunt',
 'amet',
 'tempora',
 'adipisci',
 'porro',
 'neque',
 'dolore',
 'sit',
 'labore',
 'aliquam',
 'dolor',
 'adipisci',
 'quaerat',
 'dolorem',
 'voluptatem',
 'tempora',
 'numquam',
 'velit',
 'modi',
 'numquam',
 'modi',
 'etincidunt',
 'velit',
 'quisquam',
 'quiquia',
 'sit',
 'etincidunt',
 'modi',
 'quiquia',
 

In [18]:
palabras = texto_limpio.split()

In [19]:
palabras

['numquam',
 'velit',
 'magnam',
 'quaerat',
 'numquam',
 'voluptatem',
 'voluptatem',
 'velit',
 'eius',
 'neque',
 'adipisci',
 'modi',
 'dolorem',
 'quiquia',
 'non',
 'porro',
 'est',
 'ut',
 'amet',
 'eius',
 'non',
 'etincidunt',
 'quisquam',
 'quiquia',
 'est',
 'labore',
 'sed',
 'dolorem',
 'adipisci',
 'non',
 'aliquam',
 'est',
 'modi',
 'non',
 'magnam',
 'numquam',
 'adipisci',
 'magnam',
 'neque',
 'est',
 'voluptatem',
 'modi',
 'est',
 'eius',
 'dolor',
 'porro',
 'aliquam',
 'etincidunt',
 'sit',
 'quisquam',
 'ipsum',
 'dolorem',
 'dolore',
 'ut',
 'ut',
 'dolor',
 'magnam',
 'consectetur',
 'dolor',
 'quiquia',
 'dolorem',
 'porro',
 'sed',
 'etincidunt',
 'amet',
 'tempora',
 'adipisci',
 'porro',
 'neque',
 'dolore',
 'sit',
 'labore',
 'aliquam',
 'dolor',
 'adipisci',
 'quaerat',
 'dolorem',
 'voluptatem',
 'tempora',
 'numquam',
 'velit',
 'modi',
 'numquam',
 'modi',
 'etincidunt',
 'velit',
 'quisquam',
 'quiquia',
 'sit',
 'etincidunt',
 'modi',
 'quiquia',
 

4. Vemos cuantas palabras tiene el texto 

In [20]:
len(palabras)

277

5. Para eliminar las palabras repetidas, transformamos la lista en un conjunto que no acepta repeticiones

In [21]:
palabras_unicas = set(palabras)
palabras_unicas

{'adipisci',
 'aliquam',
 'amet',
 'consectetur',
 'dolor',
 'dolore',
 'dolorem',
 'eius',
 'est',
 'etincidunt',
 'ipsum',
 'labore',
 'magnam',
 'modi',
 'neque',
 'non',
 'numquam',
 'porro',
 'quaerat',
 'quiquia',
 'quisquam',
 'sed',
 'sit',
 'tempora',
 'ut',
 'velit',
 'voluptatem'}

In [22]:
len(palabras_unicas)

27

In [23]:
palabras_unicas = list(palabras_unicas)

In [24]:
palabras_unicas

['adipisci',
 'tempora',
 'sed',
 'modi',
 'magnam',
 'est',
 'consectetur',
 'quaerat',
 'voluptatem',
 'aliquam',
 'velit',
 'dolor',
 'ipsum',
 'ut',
 'numquam',
 'sit',
 'quiquia',
 'etincidunt',
 'labore',
 'amet',
 'eius',
 'dolorem',
 'quisquam',
 'dolore',
 'porro',
 'neque',
 'non']

6. Calculamos el promedio de repeticion de las palabras 

In [26]:
round(len(palabras)/len(palabras_unicas),2)

10.26

Comprobamos que este promedio sea correcto:

In [27]:
palabras.count('dolore')

6

In [28]:
palabras.count('est')

9

In [30]:
palabras.count('ut')

7

7. Creamos un diccionario donde la clave es la palabra y el valor la cantidad de veces que se repite la palabra en el texto

Para eso primero ordenamos las palabras en orden alfabetico 

In [31]:
palabras_unicas.sort()

In [32]:
palabras_unicas

['adipisci',
 'aliquam',
 'amet',
 'consectetur',
 'dolor',
 'dolore',
 'dolorem',
 'eius',
 'est',
 'etincidunt',
 'ipsum',
 'labore',
 'magnam',
 'modi',
 'neque',
 'non',
 'numquam',
 'porro',
 'quaerat',
 'quiquia',
 'quisquam',
 'sed',
 'sit',
 'tempora',
 'ut',
 'velit',
 'voluptatem']

ahora sobre palabras_unicas ordenadas alfabeticamente creamos el diccionario 

In [33]:
ocurrencias = {}

In [34]:
palabras.count(palabras_unicas [0])

14

para no hacerlo manualmente, lo que vamos hacer es hacer un bucle 

In [35]:
for palabra in palabras_unicas:
    ocurrencias [palabra] = palabras.count(palabra)
ocurrencias

{'adipisci': 14,
 'aliquam': 10,
 'amet': 8,
 'consectetur': 10,
 'dolor': 9,
 'dolore': 6,
 'dolorem': 17,
 'eius': 13,
 'est': 9,
 'etincidunt': 14,
 'ipsum': 12,
 'labore': 6,
 'magnam': 10,
 'modi': 14,
 'neque': 7,
 'non': 18,
 'numquam': 11,
 'porro': 9,
 'quaerat': 9,
 'quiquia': 9,
 'quisquam': 13,
 'sed': 9,
 'sit': 10,
 'tempora': 3,
 'ut': 7,
 'velit': 14,
 'voluptatem': 6}

8. Ahora vemos cual es la palabra que mas se repite y cual es la palabra que menos se repite

In [36]:
ocurrencias.values()

dict_values([14, 10, 8, 10, 9, 6, 17, 13, 9, 14, 12, 6, 10, 14, 7, 18, 11, 9, 9, 9, 13, 9, 10, 3, 7, 14, 6])

In [37]:
min(ocurrencias.values())

3

In [38]:
#comprobamos
sorted(ocurrencias.values())

[3,
 6,
 6,
 6,
 7,
 7,
 8,
 9,
 9,
 9,
 9,
 9,
 9,
 10,
 10,
 10,
 10,
 11,
 12,
 13,
 13,
 14,
 14,
 14,
 14,
 17,
 18]

In [39]:
max(ocurrencias.values())

18

## Conclusion: El texto esta compuesto por un total de 277 palabras, las palabras se repiten en promedio 10,26 y la palabra que menos se repite se repite 3 veces y la que mas se repite se repite 18 veces
